Introduce IREE_USE_SYSTEM_DEPS (#20471)
openxla/iree: CMakeLists.txt

option(IREE_BUILD_PYTHON_BINDINGS "Builds the IREE python bindings" OFF)
option(IREE_BUILD_TRACY "Enables building the 'iree-tracy-capture' CLI tool and includes it in runtime Python bindings." OFF)
option(IREE_BUILD_BUNDLED_LLVM "Builds the bundled llvm-project (vs using installed)" ON)
option(IREE_USE_SYSTEM_DEPS "Uses certain system libraries instead of building downloaded dependencies" OFF)

> if (IREE_USE_SYSTEM_DEPS)

Reviewer: NIT

```suggestion
if(IREE_USE_SYSTEM_DEPS)
```

option(IREE_BUILD_PYTHON_BINDINGS "Builds the IREE python bindings" OFF)
option(IREE_BUILD_TRACY "Enables building the 'iree-tracy-capture' CLI tool and includes it in runtime Python bindings." OFF)
option(IREE_BUILD_BUNDLED_LLVM "Builds the bundled llvm-project (vs using installed)" ON)

> option(IREE_USE_SYSTEM_DEPS "Uses system libraries instead of building bundled or downloaded dependencies" OFF)

Reviewer: I think needs some additional words, including a warning  that this untested and that it is up to the user to ensure that compatible versions are available.

Author: I don't find similar words in shark-ai. Could you provide suggested changes?

Reviewer: I am perfectly fine with what @ScottTodd suggested :)

> option(IREE_USE_SYSTEM_DEPS "Uses system libraries instead of building bundled or downloaded dependencies" OFF)

Reviewer 2: How about this?

```suggestion
option(IREE_USE_SYSTEM_DEPS "Uses find_package() for certain system libraries instead of downloading. Warning: may find incompatible versions." OFF)
```

We could also specify a version [range]: https://cmake.org/cmake/help/latest/command/find_package.html.

I think it's important to say _certain_ libraries, since this new option currently only controls how `nanobind` is included. We could grow that list carefully to include other packages but there are many ways to get package discovery subtly wrong (as @marbre and I are discovering on some other projects 😅), so I'd rather start with a minimal set that you need for your specific usage.

Author: Added a cmake warning message

# Both the IREE and MLIR Python bindings require nanobind. We initialize it here
# at the top level so that everything uses ours consistently.
if(IREE_BUILD_PYTHON_BINDINGS)
  if(IREE_USE_SYSTEM_DEPS)

>     find_package(nanobind)

Reviewer: ```suggestion
    find_package(nanobind REQUIRED)
```
